### PR TITLE
Enable tagged metric names for existing Statsd metric publishing events | influxdb-statsd support

### LINKED ISSUE #11463: Tags rather than names in variable parts of the metrics
**Description**

It would be great if the metrics we publish in Airflow change to include TAGs for "variable" parts of those (such as dag_id, task_id, pool name) rather than being part of the metrics.

See here: https://airflow.apache.org/docs/1.10.12/metrics.html, https://airflow.readthedocs.io/en/latest/logging-monitoring/metrics.html  - those variable parts should be tags.

We might consider to change those metrics by default in 2.0 and possibly introduce a backport in 1.10.13 to allow people to migrate.

**Use case / motivation**

Having the variable parts of the metrics makes it really difficult to aggregate the metrics. Seems that statsd python library does not support tags, but there are possible solutions by installing extensions:

https://stackoverflow.com/questions/49852654/sending-statsd-metrics-with-tags-from-python-client

**Related Issues**

### PR DESCRIPTION
**Pull request by sungwy**
This PR attempts to close a long open issue (closes: #11463 ) related to allowing Airflow to publish tags in a backward compatible manner.

The support for tags has been long-desired, and the recent PR #28961 introduced the **tags** parameter to the StatsLogger Protocol. However, this tag only allows us to introduce tags in a forward looking way: useful existing metrics that take form of concatenated tag-like attributes cannot meaningfully be changed to be efficiently aggregated. The following are some examples of metrics that would greatly benefit from having an updated naming/tagging convention as they hold an important significance at the cluster level:

- dagrun.schedule_delay.<dag_id>
- dagrun.<dag_id>.first_task_scheduling_delay

I am introducing two additional parameters, along with one configuration property in order to enable a backward-compatible approach to introducing tags for tag-supported plugins (Datadog, Influxdb etc)

### Parameters:

- name_tags: OrderedDict[str, str]
- stat_suffix: str

**name_tags** are similar to the existing tags, except that they will be included into the end of the stat_name if **statsd_aggregation_optimized_naming_enabled** is set to False. This means that adding or changing name_tags will break backward compatibility. In contrast, the **tags** property will remain a non-breaking parameter that will allow us to introduce a breadth of tags to airflow.

**stat_suffix** are added to the stat_name after the name_tags have been concatenated. We would prefer to not have this attribute as it seems extraneous, but it is necessary to maintain backward compatibility with some metrics that have included tag-like attributes in the middle of the metric name. (e.g. dagrun.<dag_id>.first_task_scheduling_delay).

### Configuration Property

- statsd_aggregation_optimized_naming_enabled: boolean = False

When set to False (default), the current naming format is preserved. But when set to True, name_tags will be excluded from the stat_name for better aggregation. In our example, if dag_id is passed into the name_tags parameter as **OrderedDict({"dag_id": dag.dag_id})**, it will be excluded from the stat_name and instead be published as one of the tags to the metric. (The resulting stat_name will be dagrun.first_task_scheduling_delay, which allows us to aggregate the average of scheduling delays better).

--------------------------------------------------------------------------------------

### EDIT:

This PR now publishes two separate stats metrics for some important existing metrics. A single scheduling delay event will now trigger the following two Stats.incr calls. The rationale for this change is discussed in the comments.

```
Stats.timing(f"dagrun.schedule_delay.{dag.dag_id}", schedule_delay)
Stats.timing(f"dagrun.schedule_delay", schedule_delay, tags={"dag": dag.dag_id})
```

### Additional Change: InfluxDB support:

InfluxDB-StatsD support is entirely _opt-in_, and is driven by a configuration property: **statsd_influxdb_enabled** which is set to False by default. When enabled, the tags will be added into the stats_name according to the influxdb-statsd convention in order to allow the publication of tagged metrics.

I have currently made the change to **dagrun.schedule_delay.<dag_id>** or **dagrun.<dag_id>.first_task_scheduling_delay** to show how name_tags could be used to update the existing metrics in a backward-compatible way. If this sounds like a good idea, I could keep working on this PR to update other existing tags as well.

---
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Comment by syun64** [comment]
@potiuk @hussein-awala @uranusjr would love to get your thoughts on this PR

**Comment by hussein-awala** [comment]
Thank your for this PR, I was working on a PR to support the choice between adding the variables (job_name, task_id, dag_id, run_id, ...) as a part of the metric name or as tags with same metric name for DataDog metrics.

I will check your PR by tomorrow.

**Comment by syun64** [comment]
Another alternative to this PR, is just to simply publish duplicate metrics: one that has the current naming strategy, and another one that has an aggregation friendly name and its tags.

```
Stats.timing(f"dagrun.schedule_delay.{dag.dag_id}", schedule_delay)
Stats.timing(f"dagrun.schedule_delay", schedule_delay, tags={"dag": dag.dag_id})
```

**Comment by uranusjr** [comment]
> Another alternative to this PR, is just to simply publish duplicate metrics

I was thinking exactly this when I opened this PR. Is there a good way to deprecate stat names? Keeping the old names forever is probably not a very viable solution.

**Comment by syun64** [comment]
> > Another alternative to this PR, is just to simply publish duplicate metrics
> 
> I was thinking exactly this when I opened this PR. Is there a good way to deprecate stat names? Keeping the old names forever is probably not a very viable solution.

First off: thank you for the review! Very much appreciate you taking a look. I agree with you @uranusjr, that tagless metrics are very limiting for obvious reasons. But as I understand, there are a couple of factors that act against us completely deprecating the Statsd old stat names...

Firstly, it looks like the long-term vision is to deprecate Statsd support overall and move towards adopting OpenTelemetry. And a part of that proposal is to deprecate Statsd metrics / the existing stat names and introducing tagged metrics. But as I have pointed out in this issue #11463, adopting OpenTelemetry will be an organizational effort rather than an individual one for users in large companies. On top of that, I believe that that AIP is still a long way from being implemented.

Unfortunately with Statsd, although it is widely used, we have to deal with the unfortunate lack of standardization for tag-supported plugins. As an Influxdb-Statsd plugin user, I would love to just simplify the existing stat naming conventions to be more aggregation friendly and just introduce meaningful tags. But completely removing variable (dag, pool, etc) information from the existing stats metrics would be a pretty destructive change to current classic statsd users that may rely on that information.

I think if we want to reserve that destructive change until OpenTelemetry is introduced, then preserving backward compatibility as proposed in this PR might be a gentle way to bridge the gap.

**Comment by uranusjr** [comment]
If the goal is to remove statsd altogether eventually, emitting an event under multiple keys feels like the easiest solution to me.

**Comment by syun64** [comment]
> If the goal is to remove statsd altogether eventually, emitting an event under multiple keys feels like the easiest solution to me.

Sounds good to me @uranusjr . Made the changes to some important metrics. May I ask you for another review?

File: airflow/config_templates/config.yml [hunk 1]
```diff
@@ -867,6 +867,13 @@ metrics:
       type: string
       example: ~
       default: ~
+    statsd_influxdb_enabled:
+      description: |
+        To enable sending airflow metrics with StatsD-Influxdb tagging convention.
```
Comment (uranusjr) [inline airflow/config_templates/config.yml]:
```suggestion
        To enable sending Airflow metrics with StatsD-Influxdb tagging convention.
```

File: airflow/jobs/scheduler_job.py [hunk 2]
```diff
@@ -1240,7 +1240,13 @@ def _update_state(dag: DAG, dag_run: DagRun):
                 # always happening immediately after the data interval.
                 expected_start_date = dag.get_run_data_interval(dag_run).end
                 schedule_delay = dag_run.start_date - expected_start_date
+                # Publish metrics twice with backward compatible name, and then with tags
                 Stats.timing(f"dagrun.schedule_delay.{dag.dag_id}", schedule_delay)
+                Stats.timing(
+                    "dagrun.schedule_delay",
+                    schedule_delay,
+                    tags={"dag_id": f"{dag.dag_id}"},
```
Comment (uranusjr) [inline airflow/jobs/scheduler_job.py]:
```suggestion
                    tags={"dag_id": dag.dag_id},
```

File: airflow/models/dagrun.py [hunk 3]
```diff
@@ -877,6 +877,11 @@ def _emit_true_scheduling_delay_stats_for_finished_state(self, finished_tis: lis
                 true_delay = first_start_date - data_interval_end
                 if true_delay.total_seconds() > 0:
                     Stats.timing(f"dagrun.{dag.dag_id}.first_task_scheduling_delay", true_delay)
+                    Stats.timing(
+                        "dagrun.first_task_scheduling_delay",
+                        true_delay,
+                        tags={"dag_id": f"{dag.dag_id}"},
```
Comment (uranusjr) [inline airflow/models/dagrun.py]:
```suggestion
                        tags={"dag_id": dag.dag_id},
```

File: airflow/stats.py [hunk 4]
```diff
@@ -196,7 +222,23 @@ def stat_name_default_handler(stat_name, max_length=250) -> str:
...
+
+def stat_name_influxdb_handler(stat_name, max_length=250) -> str:
+    """InfluxDB-Statsd default name validator."""
+    if not isinstance(stat_name, str):
+        raise InvalidStatsNameException("The stat_name has to be a string")
+    if len(stat_name) > max_length:
+        raise InvalidStatsNameException(
+            f"The stat_name ({stat_name}) has to be less than {max_length} characters."
+        )
+    if not all((c in ALLOWED_CHARACTERS | set([",", "="])) for c in stat_name):
```
Comment (uranusjr) [inline airflow/stats.py]:
```suggestion
    if not all((c in {*ALLOWED_CHARACTERS, ",", "="}) for c in stat_name):
```

File: airflow/stats.py [hunk 5]
```diff
@@ -206,7 +248,13 @@ def stat_name_default_handler(stat_name, max_length=250) -> str:
 
 def get_current_handler_stat_name_func() -> Callable[[str], str]:
     """Get Stat Name Handler from airflow.cfg."""
-    return conf.getimport("metrics", "stat_name_handler") or stat_name_default_handler
+    handler = conf.getimport("metrics", "stat_name_handler")
+    if handler is None:
+        if conf.get("metrics", "statsd_influxdb_enabled", fallback=False):
+            handler = partial(stat_name_default_handler, allowed_chars=ALLOWED_CHARACTERS | set([",", "="]))
```
Comment (uranusjr) [inline airflow/stats.py]:
```suggestion
            handler = partial(stat_name_default_handler, allowed_chars={*ALLOWED_CHARACTERS, ",", "="})
```

File: airflow/stats.py [hunk 6]
```diff
@@ -250,43 +298,107 @@ def test(self, stat):
...
 
+def prepare_stat_with_tags(fn: T) -> T:
+    """Add tags to stat with influxdb standard format if influxdb_tags_enabled is True."""
+
+    @wraps(fn)
+    def wrapper(self, stat=None, *args, tags=None, **kwargs):
+        if self.influxdb_tags_enabled:
+            if stat is not None and tags is not None:
+                for k, v in tags.items():
+                    if not set(",=").intersection(set(v + k)):
```
Comment (uranusjr) [inline airflow/stats.py]: Probably better to use the same `all((x not in y) for x in v + k)` approach; it’s much more readable than using set here.

(Or we can introduce a util function that wraps the set operation to make it easier to understand for readers)

File: airflow/stats.py [hunk 7]
```diff
@@ -250,43 +298,107 @@ def test(self, stat):
...
     @validate_stat
-    def incr(self, stat, count=1, rate=1, tags: dict[str, str] | None = None):
+    def incr(
+        self,
+        stat,
+        count=1,
+        rate=1,
+        *,
+        tags: dict[str, str] | None = None,
+    ):
```
Comment (uranusjr) [inline airflow/stats.py]: Add types to other arguments while we’re changing these? (Same for other functions)

**Review by uranusjr — COMMENTED** [review]
Some nits. General logic lgtm.

**Comment by syun64** [comment]
> Some nits. General logic lgtm.

Thank you! I've made the suggested changes.

**Review by uranusjr — APPROVED** [review]

**Comment by syun64** [comment]
@potiuk could I ask for your review on this PR?

**Review by potiuk — APPROVED** [review]

**Comment by potiuk** [comment]
Woooho !

**Comment by syun64** [comment]
Thank you both!! @potiuk @uranusjr

**Pull request merged**